lib/SPIRV/SPIRVWriter.cpp

```diff
@@ -596,11 +596,16 @@ SPIRVFunction *LLVMToSPIRVBase::transFunctionDecl(Function *F) {
   SPIRVFunction *BF =
       static_cast<SPIRVFunction *>(mapValue(F, BM->addFunction(BFT)));
   BF->setFunctionControlMask(transFunctionControlMask(F));
-  if (F->hasName())
-    BM->setName(BF, F->getName().str());
-  if (isKernel(F))
-    BM->addEntryPoint(ExecutionModelKernel, BF->getId());
-  else if (F->getLinkage() != GlobalValue::InternalLinkage)
+  if (F->hasName()) {
+    if (isKernel(F)) {
+      /* strip the prefix as the runtime will be looking for this name */
```

Comment:
So, we get two functions with the same name in resulting SPIR-V, right? Maybe it will be better if they will have different names and the original kernel that contains body will have the prefix? I wonder if it may confuse some third party SPIR-V consumers.

Reply:
I can't figure out a way that is compatible with consumers of this library and has separate name. You can't change the non-kernel function because it breaks external linkage, and you can't change the kernel entrypoint name because it breaks runtime consumers of the translator which expect the entrypoints to have the name.

Reply:
> You can't change the non-kernel function because it breaks external linkage

I'm not following why this is the case. If you have module A declaring a kernel, and module B calling it, if you do the same name change to both, wouldn't that then link correctly?

Reply:
what happens if you generate SPIR-V from something else to link in?

Reply:
Is that a valid way to construct an OpenCL program at the API? I know we do that with the libclc helpers, but since none of those are kernels, they wouldn't be affected, would they? Unless (I didn't really read the code) this was applying a name change to all non-kernel functions, rather than just kernel function wrappers?

Reply:
I think it's legal to declare a __kernel in one file, and a non-kernel with the same name as a fwd decl in another file, and expect call to work between them, but I'd have to confirm that.

We don't rename all functions. So linkage from one file that doesn't have __kernel on it won't know to rename the underlying function.

Reply:
SPIR embeds different calling conventions for kernels vs functions. I don't think it's valid to forward-declare a function without `__kernel` and call it that way, and define it in another file with `__kernel`. That'll produce a calling convention mismatch in SPIR, and it can generate a name mismatch in SPIR-V.

Reply:
you can create multiple programs some with CreateProgramWithIL, with SPIR-V not generated from SPIRV-LLVM-Translator, you can link those against CreateProgramWithSource that do go through this. I think the linkage attributes would have to be correct for that to function properly.

Reply:
> I think it's legal to declare a __kernel in one file, and a non-kernel with the same name as a fwd decl in another file, and expect call to work between them, but I'd have to confirm that.

I can't find any info about it in the OpenCL spec where it defines __kernel attribute.
Though, I tend to agree with @jenatali , that in general it shouldn't be allowed. I've tried compiling a simple OCL example:
```
// source 1
void foo(global int *_B);     
void kernel A(global int *B) {
  foo(B);                     
}
// source 2
void kernel foo(global int *B) {
  B[0] = 1;
}                        
```
I compiled both these files to LLVM IR using `clang` and then linked them using `llvm-link`. In the resulting LLVM IR `foo` is defined as `spir_kernel`, but it is called with `spir_func` calling convention (it changes to `spir_kernel` if I add `kernel` attribute to the declaration of `foo` in source1):
```
; Function Attrs: mustprogress nofree norecurse nosync nounwind willreturn writeonly
define dso_local spir_kernel void @foo(i32 addrspace(1)* nocapture %B) local_unnamed_addr #0 !kernel_arg_addr_space !3 !kernel_arg_access_qual !4 !kernel_
arg_type !5 !kernel_arg_base_type !5 !kernel_arg_type_qual !6 {
entry:
  store i32 1, i32 addrspace(1)* %B, align 4, !tbaa !7
  ret void
}

; Function Attrs: convergent norecurse nounwind
define dso_local spir_kernel void @A(i32 addrspace(1)* %B) local_unnamed_addr #1 !kernel_arg_addr_space !3 !kernel_arg_access_qual !4 !kernel_arg_type !5
!kernel_arg_base_type !5 !kernel_arg_type_qual !6 {
entry:
  call spir_func void @foo(i32 addrspace(1)* %B) #2
  ret void
}
```
Even though LLVM verifier doesn't complain about this, LLVM langref says that:

> The calling convention of the call must match the calling convention of the target function, or else the behavior is undefined.

I don't think that UB case should be somehow allowed or taken into account.
It seems unfortunately this can't be diagnosed by front-ends but maybe it can be reported by linker.
From the other hand SPIR-V spec says:
```
The Linkage Attributes Decoration must not be applied to functions targeted by an OpEntryPoint instruction.
```
After that I wonder if it is meant to call a kernel from a separate translation unit at all. (@bashbaug , I'd appreciate if you could help here).

Reply:
See https://github.com/KhronosGroup/SPIRV-Registry/issues/74: This is required by OpenCL, but disallowed by SPIR-V, which is the whole point of this change.

The only question in my mind at this point is, can you take one `clProgram` which is source (created from `CreateProgramWithSource`), and expect to link it with another created with `CreateProgramWithIL`. I'd argue no, but I don't see anything explicitly saying yes or no in the CL spec. That probably needs a separate issue on https://github.com/KhronosGroup/OpenCL-Registry.

Reply:
> After that I wonder if it is meant to call a kernel from a separate translation unit at all.

Good question.  I'm guessing this is a corner case nobody has encountered / considered before.  I can see how this would be complicated given the SPIR-V constraints.  Should this scenario be dis-allowed?  Would that enable this PR to move forward?

> The only question in my mind at this point is, can you take one clProgram which is source (created from CreateProgramWithSource), and expect to link it with another created with CreateProgramWithIL. I'd argue no, but I don't see anything explicitly saying yes or no in the CL spec.

This is interesting too, because I'd argue that this should be supported.  😄 

Rationale is: the program object created from source or IL must be compiled (via `clCompileProgram`) before it can be linked, which transitions the state of the program object from `BINARY_TYPE_NONE` to `BINARY_TYPE_COMPILED_OBJECT`.  After the program object has been compiled I can't think of any reason to differentiate between program objects that originated from source and program objects that originate from IL, so the linking should "just work".  Is there a detail I'm missing here?

I agree both of these topics should have separate issues for more discussion but they should be filed in in [OpenCL-Docs](https://github.com/KhronosGroup/OpenCL-Docs/), not [OpenCL-Registry](https://github.com/KhronosGroup/OpenCL-Registry).

Reply:
> After the program object has been compiled I can't think of any reason to differentiate between program objects that originated from source and program objects that originate from IL, so the linking should "just work". Is there a detail I'm missing here?

C++ compilers have similar problems here, where a "compiled object" from one version of a compiler doesn't necessarily have the same ABI as a "compiled object" from another version of a compiler, and therefore the two objects can't necessarily be linked. This is similar, if we treat an IL program as an object from one version of a compiler, which might not be the same as the platform's source compiler.

> Should this scenario be dis-allowed? Would that enable this PR to move forward?

I don't think this is strictly necessary, what airlied has here now (two functions with the same name, one exported, one as a kernel) is probably fine, and might even be desirable since it round-trips better. While writing this comment, I've come around to this POV, but I still think it's good to probe on what the expected behavior here is - and if it is to be explicitly required, it absolutely should be tested by the CTS.

Reply:
I'm pretty sure it's allowed to have IL and non-IL mixed in one path. So I think this should remain as-is.

Reply:
Just FYI, I've created two OpenCL-Docs issues based on the above discussion:

* https://github.com/KhronosGroup/OpenCL-Docs/issues/675
* https://github.com/KhronosGroup/OpenCL-Docs/issues/676